Add Private key jwt python blog post (#918)
oktadeveloper/okta-blog: _source/_posts/2021-10-08-private-key-jwt-python.md

---
layout: blog_post
title: "Set Up the Private Key JWT Flow in Three Python Commands"
author: cale-switzer

> by: advocate

Reviewer: ```suggestion
by: internal-contributor
```
this 'role' is for staff who aren't on the dev-advocacy team

communities: [python]
description: "The Private Key JWT flow is one of the more complicated OAuth flows to set up. Learn how the flow works at a high level and how you can implement it in three commands with Python scripting."
tags: [python, oauth, private-key-jwt]
image:
type: awareness
---

The [Private Key JWT flow](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) is one of the more complicated OIDC flows to set up. If you already know HOW it works and just want to get to the code, [click here](#python-private-key-jwt-setup) to skip ahead. If you're looking to gain a better understanding of how it works, here's a high level explanation:

> {% img blog/private-key-jwt-python/pkj_1.png alt:"Private Key JWT FLowchart" width:"1000" %}{: .center-image }

Reviewer 2: summary at the end, "You can now use this token to perform management tasks on your Okta org based on your defined scopes by passing it in the **Authorization** header as **Bearer {access_token}**.", was what made the motivation behind this article click for me. Would be good to include something in the intro about "if you need a token to perform management tasks on your okta org" or whatever, to clarify what problem the code is solving.

Author: Good thought, I changed the intro to:

The [Private Key JWT flow](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) is one of the more complicated OIDC flows to set up. At Okta, it's commonly used to get OAuth4Okta access tokens which enable you to perform management tasks on your org. If you already know HOW it works and just want to get to the code, [click here](#python-private-key-jwt-setup) to skip ahead. If you're looking to gain a better understanding of how it works, here's a high level explanation:

Once you've obtained the access token from this flow, you can throw it into the Authorization header of requests to your Okta APIs:

{% img blog/private-key-jwt-python/pkj_2.png alt:"Use Access Token" width:"400" %}{: .center-image }

But let's break it down a bit more:

> ### Step 1

Reviewer 2: ```suggestion
### Step 1: Generate keypair
```
we switched from title case to sentence case for blog post section headings early in Oct. to be consistent with Dev Docs style and Okta InfoDev style, and my own preference. The blog title is still in sentence case.

> Typically, one may think of JWTs as strictly being in the realm of the Identity Provider (IdP) - and in most cases that's correct. When an IdP grants tokens to a service provider (SP), it is signing those tokens with a private/secret key. Part of the responsibility of the SP when validating tokens is to use a public key provided by the IdP to verify the signature.

Reviewer: ```suggestion
Typically, one may think of JWTs as strictly being in the realm of the *identity provider* (IdP) - and in most cases that's correct. When an IdP grants tokens to a service provider (SP), it is signing those tokens with a private/secret key. Part of the responsibility of the SP when validating tokens is to use a public key provided by the IdP to verify the signature.
```
using italics for first mention of "identity provider" instead of uppercase

The Private Key JWT flow flips this on its head - at least at the start. It's now the SP's turn to generate a public/private keypair, share the public key with the IdP (more on that in Step 2), create a JWT and sign it with their private key (more on that in Step 3).

But I digress...Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048 bit key size).

> ### Step 2

Reviewer 2: ```suggestion
### Step 2: Register public key
```
switching to sentence case

> The next step is registering your public key with the IdP - with Okta, that's done through our [clients](https://developer.okta.com/docs/reference/api/oauth-clients/#register-new-client) API. With this step, Okta will create an application integration and generate a unique client ID - this value will be important when it comes time to create the Private Key JWT. It will be this client ID as well as the working public key decryption that validates your request.

Reviewer 2: passive voice on the "that's done through". Did you mean "the script does it through our clients API"?

Reviewer: ```suggestion
The next step is registering your public key with the IdP. With Okta, that's done through our [clients](https://developer.okta.com/docs/reference/api/oauth-clients/#register-new-client) API. With this step, Okta will create an application integration and generate a unique client ID - this value will be important when it comes time to create the Private Key JWT. It will be this client ID as well as the working public key decryption that validates your request.
```
Made one long-ish compound sentence into two sentences.

In other flows like the popular [Authorization Code](https://developer.okta.com/docs/guides/implement-grant-type/authcode/main) flow, the identity provider generates a secret value called the client secret. In the Private Key JWT flow, the burden is on the service provider to generate a secret value for the same purpose - albeit by different means.

> ### Step 3

Reviewer 2: ```suggestion
### Step 3: Get access token
```

> Technically this could be broken down into two separate steps - creating the private key JWT and then submitting it to the IdP to obtain an access token. I lumped it into one step because the Python command `get_access_token.py` below does both of these at once.

Reviewer 2: "technically" sounds like a "well actually" here. How about "There are two components to this step, but the Python command does them together so we can treat them as a single operation" or similar?

> Using the private key created in step 1, you sign a JWT with the following claims:

Reviewer 2: I keep having to look up what JWT claims are whenever I see it. can `claims` here be a link to those docs, for those similarly ignorant?

Author: done!

```
'iss': {client ID generated in step 2},
'sub': {client ID generated in step 2},
'aud': {okta_url}/oauth2/v1/token,
"exp": {epoch expiration at future time}
```

This token is sent to the /token endpoint of the IdP:

```
  POST /token HTTP/1.1
  Host: server.example.com
  Content-Type: application/x-www-form-urlencoded

  grant_type=client_credentials&
    scope={scopes}&
    client_id={client ID generated in step 2}&
    client_assertion_type=
    urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&
    client_assertion={private key JWT}
```

> Assuming everyting is setup correctly, the IdP should respond with an access token that you can use to access scoped resources.

Reviewer: ```suggestion
Assuming everything is set up correctly, the IdP should respond with an access token that you can use to access scoped resources.
```
spelling typo, also set up as verb is two words.

> For an even deeper look at this flow, you can check out our walkthrough [here](https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/overview/). For now, it's time to set this up with Python!

Reviewer: ```suggestion
For an even deeper look at this flow, you can [check out our walkthrough](https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/overview/). For now, it's time to set this up with Python!
```
Style guide best practice: We avoid use of "here" "click this" etc - instead use a word or short phrase that describes more or less where the link will go.

## Python Private Key JWT Setup

> 1. Clone the repo found [here](https://github.com/Tennyx/private-key-jwt-python), open terminal and `cd` into the project root.

Reviewer: ```suggestion
1. Clone the [OktaDev Private Key repo](https://github.com/Tennyx/private-key-jwt-python), open terminal, and `cd` into the project root.
```
moved the link, added serial comma

> 2. Create a new [virtualenv](https://docs.python.org/3/library/venv.html) in the root folder with `virtualenv env`

Reviewer 2: in py3 this is idiomatically invoked at `python3 -m venv env` -- standalone `virtualenv` to create one sadly seems to have died with py2 unless you set up aliases for it

Reviewer: ```suggestion
2. Create a new [virtualenv](https://docs.python.org/3/library/venv.html) in the root folder with `virtualenv env`.
```
add period

3. Run the virtual environment with `source env/bin/activate`
4. Install necessary Python packages with `pip install -r requirements.txt`

> 5. In the root folder, create a new .env file and paste the following values into it:

Reviewer 2: ```suggestion
5. In the root folder, create a new `.env` file and paste the following values into it:
```

```
PRIVATE_KEY=
PUBLIC_KEY=
MODULUS=
CLIENT_ID=
OKTA_URL=
API_KEY=
SCOPES=
```

> Scripts will generate values for the first 4 vars. You will need to fill out the other values. **Note:** the API key is needed for creating the OAuth client in Okta, so make sure the token you use has permissions to create applications.

Reviewer 2: How? Where does one get these? docs links to how to find each would be very helpful. The examples given for the scopes are helpful.

Reviewer 2: also, you say the scripts will generate the first 4, but after i ran generate_keys.py it filled out the private and public keys, removed the modulus (????), and did not fill out CLIENT_ID (surprising), or OKTA_URL, API_KEY, or SCOPES (expected based on docs

Author: can you double check the MODULUS var? I just ran it with a fresh .env and it populated. It can blend in with PUBLIC_KEY above it.

Also, CLIENT_ID doesn't get generated until the 2nd command after an Okta app is made

Author: Added more clarity to this

- OKTA_URL=https://{your_okta_domain}
- API_KEY={okta_api_key}
- SCOPES='okta.users.read okta.users.manage'

*The SCOPES var is an string of scopes (seperated by spaces) granted to your eventual access token that you will be using for management tasks. I put in a couple common ones as a placeholder. In your Okta application, it would look like this:*

> ![SCOPES_1](https://i.imgur.com/B4zNpom.png)

Reviewer 2: question for @havi - do we trust imgur to keep hosting stuff as long as we want it to, or should the image be dropped in the blog repo somewhere?

Reviewer: The image should move to the images 'folder' associated with the blog post. Thanks both!

Author: good catch, forgot to change that when copying over from my GH repo!

```
///////////////// Okta Service app created: 

{application object}

///////////////// Setting CLIENT_ID in .env:

{client ID of newly created app}

///////////////// Scopes granted: 

{scopes granted to the application}
```

If you navigate to your applications in Okta, you will see this newly created app under the name **Private Key JWT Service App**.

### Step 3: Create PKJ and Get Access Token

The above two scripts should only need to be ran once for setup. From hereon out, you can run the following command to get an access token from Okta:

```
python get_access_token.py
```

You should see the access token printed to terminal:

```
///////////////// Access Token:

{access_token}
```

You can now use this token to perform management tasks on your Okta org based on your defined scopes by passing it in the **Authorization** header as **Bearer {access_token}**.

## In Conclusion

The Private Key JWT flow is complicated to set up, but once the pieces are in place it should be pretty automated. As always, the private key is the key to the castle. So...

> ![Keep it secret, keep it safe!](https://thumbs.gfycat.com/HonestSmartBuffalo-max-1mb.gif)

Reviewer 2: for accessibility, probably best to caption this image with "keep it secret, keep it safe!" as well?

Reviewer: One additional comment for the conclusion. @Tennyx all Okta Dev blog posts close with a call to action that includes links to where the reader can go to follow us (twitter, youtube, etc). Please grab this copy from a blog post of your choice, and add it in your own words. You can see the one we added on your previous post - https://developer.okta.com/blog/2021/09/01/flask-scim-server and grab that...

Reviewer 2: ```suggestion
![Keep it secret, keep it safe!](https://thumbs.gfycat.com/HonestSmartBuffalo-max-1mb.gif)

Got questions? You can leave them in the comments below! Want to stay in touch? Follow our social channels: [@oktadev on Twitter](https://twitter.com/oktadev), [Okta for Developers on LinkedIn](https://www.linkedin.com/company/oktadev), [Twitch](https://www.twitch.tv/oktadev), and [YouTube](https://youtube.com/oktadev).

```

image:
type: awareness
---

The [Private Key JWT flow](https://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) is one of the more complicated OIDC flows to set up. If you already know HOW it works and just want to get to the code, [click here](#python-private-key-jwt-setup) to skip ahead. If you're looking to gain a better understanding of how it works, here's a high level explanation:

{% img blog/private-key-jwt-python/pkj_1.png alt:"Private Key JWT FLowchart" width:"1000" %}{: .center-image }

Once you've obtained the access token from this flow, you can throw it into the Authorization header of requests to your Okta APIs:

{% img blog/private-key-jwt-python/pkj_2.png alt:"Use Access Token" width:"400" %}{: .center-image }

But let's break it down a bit more:

### Step 1: Generate keypair

Typically, one may think of JWTs as strictly being in the realm of the *identity provider* (IdP) - and in most cases that's correct. When an IdP grants tokens to a service provider (SP), it is signing those tokens with a private/secret key. Part of the responsibility of the SP when validating tokens is to use a public key provided by the IdP to verify the signature.

> The Private Key JWT flow flips this on its head - at least at the start. It's now the SP's turn to generate a public/private keypair, share the public key with the IdP (more on that in Step 2), create a JWT and sign it with their private key (more on that in Step 3).

Reviewer: ```suggestion
The Private Key JWT flow flips this on its head - at least at the start. It's now the SP's turn to generate a public/private keypair, share the public key with the IdP (more on that in Step 2), create a JWT, and sign it with their private key (more on that in Step 3).
```
fussy serial comma

> But I digress...Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048 bit key size).

Reviewer: ```suggestion
But I digress...Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048-bit key size).
```
hyphen

> But I digress...Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048-bit key size).

Reviewer: ```suggestion
But I digress... Step 1 is really just about generating the keypair. (In this case, RSA keys with a 2048-bit key size.)
```

Using the private key created in step 1, you sign a JWT with the following claims:

```
'iss': {client ID generated in step 2},
'sub': {client ID generated in step 2},
'aud': {okta_url}/oauth2/v1/token,
"exp": {epoch expiration at future time}
```

This token is sent to the /token endpoint of the IdP:

```
  POST /token HTTP/1.1
  Host: server.example.com
  Content-Type: application/x-www-form-urlencoded

  grant_type=client_credentials&
    scope={scopes}&
    client_id={client ID generated in step 2}&
    client_assertion_type=
    urn%3Aietf%3Aparams%3Aoauth%3Aclient-assertion-type%3Ajwt-bearer&
    client_assertion={private key JWT}
```

Assuming everything is set up correctly, the IdP should respond with an access token that you can use to access scoped resources.

For an even deeper look at this flow, you can [check out our walkthrough](https://developer.okta.com/docs/guides/implement-oauth-for-okta-serviceapp/overview/). For now, it's time to set this up with Python!

> ## Python Private Key JWT Setup

Reviewer: ```suggestion
## Python private key JWT setup
```
we've switched to sentence case headers

1. Clone the [OktaDev Private Key repo](https://github.com/Tennyx/private-key-jwt-python), open terminal, and `cd` into the project root.
2. Create a new [virtualenv](https://docs.python.org/3/library/venv.html) in the root folder with `virtualenv env`.

> 3. Run the virtual environment with `source env/bin/activate`

Reviewer: ```suggestion
3. Run the virtual environment with `source env/bin/activate`.
```
consistent punctuation

> 4. Install necessary Python packages with `pip install -r requirements.txt`

Reviewer: ```suggestion
4. Install necessary Python packages with `pip install -r requirements.txt`.
```
punctuation

5. In the root folder, create a new `.env` file and paste the following values into it:

```
PRIVATE_KEY=
PUBLIC_KEY=
MODULUS=
CLIENT_ID=
OKTA_URL=
API_KEY=
SCOPES=
```

Scripts will generate values for the first 4 vars. You will need to fill out the other values. **Note:** the API key is needed for creating the OAuth client in Okta, so make sure the token you use has permissions to create applications.

- OKTA_URL=https://{your_okta_domain}
- API_KEY={okta_api_key}
- SCOPES='okta.users.read okta.users.manage'

*The SCOPES var is an string of scopes (seperated by spaces) granted to your eventual access token that you will be using for management tasks. I put in a couple common ones as a placeholder. In your Okta application, it would look like this:*

![SCOPES_1](https://i.imgur.com/B4zNpom.png)

> ## Running the Scripts

Reviewer: ```suggestion
## Running the scripts
```
sentence case headers

To run these scripts, make sure you are still in the root folder of the cloned repository.

> ### Step 1: Generate Keypair

Reviewer: ```suggestion
### Step 1: Generate keypair
```
sentence case headers

Run the following command in terminal:

```
python generate_keys.py
```

You should see confirmation printed to the terminal:

```
///////////////// Keys generated and added to .env file.
```

Sure enough, you can pop over to the .env file to see the `PRIVATE_KEY`, `PUBLIC_KEY` and `MODULUS` values populated there.

> ### Step 2: Create OAuth Service App in Okta

Reviewer: ```suggestion
### Step 2: Create OAuth service app in Okta
```
sentence case headers

Now that you have the keypair, you can create the OAuth app in Okta with the public key. The scopes you defined in the .env file will also be granted in the application. Type the following into the terminal:

```
python create_okta_service_app.py
```

You will get confirmation in the terminal of the created app, that the client ID of the app was updated in your .env file, as well as scopes granted.

```
///////////////// Okta Service app created: 

{application object}

///////////////// Setting CLIENT_ID in .env:

{client ID of newly created app}

///////////////// Scopes granted: 

{scopes granted to the application}
```

If you navigate to your applications in Okta, you will see this newly created app under the name **Private Key JWT Service App**.

> ### Step 3: Create PKJ and Get Access Token

Reviewer: ```suggestion
### Step 3: Create PKJ and get access token
```
sentence case headers

> The above two scripts should only need to be ran once for setup. From hereon out, you can run the following command to get an access token from Okta:

Reviewer: ```suggestion
The two scripts above should only need to run once for setup. From hereon out, you can run the following command to get an access token from Okta:
```
verb tense tweak

```
python get_access_token.py
```

You should see the access token printed to terminal:

```
///////////////// Access Token:

{access_token}
```

You can now use this token to perform management tasks on your Okta org based on your defined scopes by passing it in the **Authorization** header as **Bearer {access_token}**.

> ## In Conclusion

Reviewer: ```suggestion
## In conclusion: Keep it secret, keep it safe
```
adds context/info to the heading.

The Private Key JWT flow is complicated to set up, but once the pieces are in place it should be pretty automated. As always, the private key is the key to the castle. So...

> ![Keep it secret, keep it safe!](https://thumbs.gfycat.com/HonestSmartBuffalo-max-1mb.gif)

Reviewer: The animated gif linked here is broken in preview mode. Not sure how to verify or resolve.

Reviewer 3: Looks like it's working now 😄
